src/clj/ataru/applications/application_store.clj

```diff
@@ -501,3 +501,42 @@
                                                                    :hakemus_oids    [""]})
        (map unwrap-person-and-hakemus-oid)
        (into {})))
+
+(defn- assert-correct-from-state-for-review
+  [application-review from-state]
+  (assert (or (= (:state application-review) from-state)
+              (and (nil? (:state application-review))
+                   (= from-state (ffirst ataru.application.review-states/application-review-states))))))
+
+(defn mass-update-application-states
+  [session application-keys from-state to-state]
+  (jdbc/with-db-transaction [conn {:datasource (db/get-datasource :db)}]
+    (let [connection       {:connection conn}
+          username         (get-in session [:identity :username])
+          organization-oid (get-in session [:identity :organizations 0 :oid])]
+      (doseq [application-key application-keys]
+        (let [existing-review   (get-application-review application-key)
+              new-review        (if existing-review
+                                  (transform-keys ->snake_case
+                                                  (-> existing-review
+                                                      (assoc :state to-state)
+                                                      (dissoc :modified-time)))
+                                  {:state to-state
+                                   :application_key application-key})
+              application-event {:application_key  application-key
+                                 :event_type       "review-state-change"
+                                 :new_review_state to-state
+                                 :virkailija_oid   nil
+                                 :hakukohde        nil
+                                 :review_key       nil}]
+          (assert-correct-from-state-for-review existing-review from-state)
+          (if existing-review
+            (yesql-save-application-review! new-review connection)
+            (yesql-add-application-review! new-review connection))
+          (yesql-add-application-event!
+            application-event
+            connection)
+          (audit-log/log {:new              application-event
```

Comment:
Auditlogitus pitäis tehdä transaktion ulkopuolella kun tiedetään, että kaikki meni putkeen.